enclave/core/printf.c

```diff
@@ -0,0 +1,712 @@
+        out->str[out->off + n] = '\0';
+    }
+
+    out->off += count;
+
+    return count;
+}
+
+static void _oe_out_str_init(oe_out_str_t* out, char* str, size_t size)
+{
+    out->base.write = _write;
+    out->str = str;
+    out->size = size;
+    out->off = 0;
+}
+
+/*
+**==============================================================================
+**
+** Public definitions:
+**
+**==============================================================================
+*/
+
+//
+// Produce output according to a given format string.
+//
+// This function is similar to vsnprintf() but has no support for floating
+// point types.
+//
+// @param str Write output to this string
+// @param size The size of **str** parameter.
+// @param fmt The limited printf style format.
+//
+// @returns The number of characters that would be written excluding the
+// zero-terminator. If this value is greater or equal to **size**, then the
+// string was truncated.
+//
+//
+int oe_vsnprintf(char* str, size_t size, const char* fmt, oe_va_list ap)
+{
+    oe_out_str_t out;
+
+    if (!str && size != 0)
+        return -1;
+
+    _oe_out_str_init(&out, str, size);
+
+    return _vprintf(&out.base, fmt, ap);
+}
+
+//
+// Produce output according to a given format string.
+//
+// This function is similar to snprintf() but has limited support for format
+// types. See oe_vsnprintf() for details on these limits.
+//
+// @param str Write output to this string.
+// @param size The size of **str** parameter.
+// @param fmt The limited printf style format.
+//
+// @returns The number of characters that would be written excluding the
+// zero-terminator. If this value is greater or equal to **size**, then the
+// string was truncated.
+//
+int oe_snprintf(char* str, size_t size, const char* fmt, ...)
+{
+    oe_va_list ap;
+    oe_va_start(ap, fmt);
+    int n = oe_vsnprintf(str, size, fmt, ap);
+    oe_va_end(ap);
+    return n;
+}
+
+int oe_vprintf(const char* fmt, oe_va_list ap_)
+{
+    char buf[256];
+    char* p = buf;
+    int n;
+    char* new_buf = NULL;
+
+    /* Try first with a fixed-length scratch buffer */
+    {
+        oe_va_list ap;
+        oe_va_copy(ap, ap_);
+        n = oe_vsnprintf(buf, sizeof(buf), fmt, ap);
+        oe_va_end(ap);
+
+        if (n < 0)
+            goto done;
+
+        if ((size_t)n < sizeof(buf))
+        {
+            oe_host_write(0, p, (size_t)-1);
+            goto done;
+        }
+    }
+
+    /* If string was truncated, retry with correctly sized buffer */
+    {
+        if (!(new_buf = (char*)oe_malloc((size_t)n + 1)))
+            goto done;
+
+        p = new_buf;
+
+        oe_va_list ap;
+        oe_va_copy(ap, ap_);
+        n = oe_vsnprintf(p, (size_t)n + 1, fmt, ap);
+        oe_va_end(ap);
+
+        if (n < 0)
+            goto done;
+
+        oe_host_write(0, p, (size_t)-1);
+    }
+
+done:
+
+    if (new_buf)
+        oe_free(new_buf);
+
+    return n;
+}
+
+int oe_printf(const char* format, ...)
```

Comment:
This is a lot of code to reimplement printf.
If an enclave needs printing, it can link against libc. I would recommend getting rid of printf from oe_core.

Reply:
We use oe_printf() to add debugging statements between oe_encave to oe_core. These modules cannot link libc. If we remove it, there will be no way to add any debug printf's to the enclave stack. We are moving libc out of the OE core (which includes crypto, mbedtls, etc.).

Reply:
Given that this is for debugging and that we support the standard printf, we could make oe_printf a function pointer that needs to be setup if debug-printing is needed.
We rarely create any enclaves that don't link against libc; and it seems like extra code we need to maintain by having our own version of printf.

Reply:
This PR introduces `oe_vprintf` & `oe_printf` which are largely wrappers over the existing `oe_vsnprintf` and `oe_snprintf` methods in the removed snprintf.c file. The elibc implementation is more robust and fixes a couple of existing bugs in the old `oe_snprintf`, and represents the bulk of the code in the source file.

There are methods in ctype.c and strtoul.c that take arguably non-essential dependencies on this today (for bootstrapping generation of the static character tables), so the PR left this in for completeness. I'm open to cutting out the two additional methods, but other enclave code already depends on `oe_vsnprintf` for logging etc. so that's a larger refactor.